### DISCUSSION TITLE
Export as PDF/A

### DISCUSSION THREAD

**Discussion by bennetrr** [mindee/doctr#869] [Q&A] [answered]
Hello,
I've seen an issue here about how to export the OCR as a PDF/A-2B file, but I couldn't find anything in the documentation.

Is this possible with your library?

**Comment by felixdittrich92** [comment, marked as answer]
Hi @bennetrr :wave: ,
Yes it is :) take a look into the notebooks section (documentation) -> Link to 
[Export PDF/A](https://github.com/mindee/notebooks/blob/main/doctr/export_as_pdfa.ipynb)

[Colab notebook](https://colab.research.google.com/github/mindee/notebooks/blob/main/doctr/export_as_pdfa.ipynb)  -> update first pip please to: `!pip3 install python-doctr==0.5.1`  (ping @charlesmindee post release modifications)

Two things to note: 
- with the last release there was a change in `DocumentFile.from_pdf(..).as_images()` to `DocumentFile.from_pdf(..)` we need to update this in the notebook :+1: 
- there are currently only straight boxes supported to export as PDF/A so you can pass `export_as_straight_boxes=True` to the predictor if your documents contain rotated text 

Does this answer your question ?
Let me know if you have further questions :smiley:

**Reply by bennetrr** [reply]
I will try this when I'm at home, but I have one more question:

Is there any plan to support this directly in the library? Like a function `doc.save_as_pdfa()`

**Reply by felixdittrich92** [reply]
@bennetrr 
Would be great if you can open an issue for this feature request :hugs: 

@fg-mindee normally the integration wouldn't be a problem  if we can make friends with `reportlab` (it's lightweight without 3rd party dependencies - BSD license) !? :joy:

**Reply by bennetrr** [reply]
So far it works, now I wait for what the others say about it

I'm also creating the issue

**Reply by bennetrr** [reply]
The result looks fine, but there are no spaces when we are copying the text. Is this intended or did we forget something?

**Reply by felixdittrich92** [reply]
@bennetrr 
Good catch :+1:  I have opened a PR to update the notebook https://github.com/mindee/notebooks/pull/10

ftm you can replace the `export_pdfa` method with following: ( if have marked the new lines with # <----- NEW)

```
def export_pdfa(self,
                    out_filename: str,
                    hocr: ET.ElementTree,
                    image: Optional[np.ndarray] = None,
                    fontname: str = "Times-Roman",
                    fontsize: int = 12,
                    invisible_text: bool = True,
                    add_spaces: bool = True,   # <--- NEW
                    dpi: int = 300):
        """
        Generates a PDF/A document from a hOCR document.
        """

        width, height = None, None
        # Get the image dimensions
        for div in hocr.findall(".//div[@class='ocr_page']"):
            coords = self._element_coordinates(div)
            pt_coords = self._pt_from_pixel(coords, dpi)
            width, height = pt_coords['x2'] - \
                pt_coords['x1'], pt_coords['y2'] - pt_coords['y1']
            # after catch break loop
            break
        if width is None or height is None:
            raise ValueError("Could not determine page size")

        pdf = Canvas(out_filename, pagesize=(width, height), pageCompression=1)

        span_elements = [element for element in hocr.iterfind(".//span")]
        for line in span_elements:
            if 'class' in line.attrib and line.attrib['class'] == 'ocr_line' and line is not None:
                # get information from xml
                pxl_line_coords = self._element_coordinates(line)
                line_box = self._pt_from_pixel(pxl_line_coords, dpi)

                # compute baseline
                slope, pxl_intercept = self._get_baseline(line)
                if abs(slope) < 0.005:
                    slope = 0.0
                angle = atan(slope)
                cos_a, sin_a = cos(angle), sin(angle)
                intercept = pxl_intercept / dpi * inch
                baseline_y2 = height - (line_box['y2'] + intercept)

                # configure options
                text = pdf.beginText()
                text.setFont(fontname, fontsize)
                pdf.setFillColor(black)
                if invisible_text:
                    text.setTextRenderMode(3)  # invisible text

                # transform overlayed text
                text.setTextTransform(
                    cos_a, -sin_a, sin_a, cos_a, line_box['x1'], baseline_y2)

                elements = line.findall(".//span[@class='ocrx_word']")
                for elem in elements:
                    elemtxt = self._get_element_text(elem).strip()
                    # replace unsupported characters
                    elemtxt = elemtxt.translate(str.maketrans(
                        {'ﬀ': 'ff', 'ﬃ': 'f‌f‌i', 'ﬄ': 'f‌f‌l', 'ﬁ': 'fi', 'ﬂ': 'fl'}))
                    if not elemtxt:
                        continue

                    # compute string width
                    pxl_coords = self._element_coordinates(elem)
                    box = self._pt_from_pixel(pxl_coords, dpi)
                    # <---- START NEW
                    if add_spaces: 
                        elemtxt += ' '
                        box_width = box['x2'] + \
                            pdf.stringWidth(elemtxt, fontname,
                                            fontsize) - box['x1']
                    else:
                        box_width = box['x2'] - box['x1']
                    # <---- END NEW
                    font_width = pdf.stringWidth(elemtxt, fontname, fontsize)

                    # Adjust relative position of cursor
                    cursor = text.getStartOfLine()
                    dx = box['x1'] - cursor[0]
                    dy = baseline_y2 - cursor[1]
                    text.moveCursor(dx, dy)

                    # suppress text if it is 0 units wide
                    if font_width > 0:
                        text.setHorizScale(100 * box_width / font_width)
                        text.textOut(elemtxt)
                pdf.drawText(text)

        # overlay image if provided
        if image is not None:
            pdf.drawImage(ImageReader(Image.fromarray(image)),
                          0, 0, width=width, height=height)
        pdf.save()
```

**Reply by felixdittrich92** [reply]
@bennetrr has this solved your issue ? :)

**Reply by bennetrr** [reply]
No, not really. Even with the `add_spaces` option there are no spaces between the words. And the text that is written into the PDF usually does not match the actual size.

I searched a while for other hOCR to PDF libraries, but I couldn't find anything. Then I remembered that OCRmyPDF (which we used before for the OCR) has also to write the hOCR into the PDF, and now I'm using their API directly and it works really good. If you want I can post the code here.

EDIT: Seems that most of the text still does not contain spaces

**Reply by felixdittrich92** [reply]
Hi @bennetrr ,
thanks for your feedback :)
I'm a bit confused i have tested also some examples and on my side it works well, would it be possible to get 1 or 2 examples where it's not working ? :) Thanks again 👋

**Reply by bennetrr** [reply]
We can give you the PDF where the problems occur, but it takes some time because we have to ask the owners first. We deleted all information, so they should be fine I think

**Reply by felixdittrich92** [reply]
Thanks this would be really nice :)

**Comment by bennetrr** [comment]
Here are the files:
[censored.pdf](https://github.com/mindee/doctr/files/8441816/censored.pdf)
[p1.hocr.txt](https://github.com/mindee/doctr/files/8441818/p1.hocr.txt)

(I had to rename the second file because GitHub only accepts certain file types)

**Comment by felixdittrich92** [comment]
@bennetrr 
Thanks a lot :+1:
